### DISCUSSION TITLE
How to invoke a function that requires an Enum?

### DISCUSSION THREAD

**Discussion by SamueleD78** [vfsfitvnm/frida-il2cpp-bridge#145] [Q&A]
First, thank you for this WONDERFUL tool, i really LOVE it ^_^

I was able to do almost everything, but i'm having serious troubles in managing the enums.
I saw there was an older post, but has been deleted :-(

The function i wanna invoke is:
`Game.ItemAssetInfo GetAssetInfo(Game.ItemDefine.EItemType itemType); // 0x0319f924
`
the enum Game.ItemDefine.EItemType is (just a slice):

```
enum Game.ItemDefine.EItemType : System.Enum
{
    System.Byte value__; // 0x10
    static Game.ItemDefine.EItemType None = 0;
    static Game.ItemDefine.EItemType KouraG = 1;
    static Game.ItemDefine.EItemType KouraR = 2;
}

```
I'm trying everything to create an instance of the enum and to invoke the function, but without reason, i always get a: "TypeError: cannot read property 'message' of null"

I swear i've read everything on the wiki and searched the web but without result.
Please help ^^

**Comment by vfsfitvnm** [comment]
> I saw there was an older post, but has been deleted 

I apologize, I moved some issues to this section, so every link is now invalid.

However, this is what you have to do:
```ts
const EItemType: Il2Cpp.Class = ...;
const KouraG = EItemType.field<Il2Cpp.ValueType>("KouraG").value;
```
Then you can pass `KouraG` to the function you desire to call.

**Reply by SamueleD78** [reply]
first and foremost, thank you so much for your very fast answer ^^
I would had never figured it out by myself.

I've tryed it right away, and the result is: Error: abort was called

In the same class, there is another function returning that same Enum, and it tryed passing the result of that function too.
same output.

```
const EItemType: Il2Cpp.Class = mscorlib.classes[3548];
const KouraG = EItemType.field<Il2Cpp.ValueType>("KouraG").value;
console.log(KouraG); // perfect!
console.log(ParamAssetCatalog.method<Il2Cpp.Pointer>("GetAssetInfo").invoke(KouraG)); // Error: abort was called

const test = new Il2Cpp.Array<Il2Cpp.Object>(ParamAssetCatalog.method<Il2Cpp.Pointer>("get_ValidItems").invoke());
console.log(test.get(0)); // perfect, same as above
console.log(ParamAssetCatalog.method<Il2Cpp.Pointer>("GetAssetInfo").invoke(ggg.get(0)));  // Error: abort was called

```

**Reply by vfsfitvnm** [reply]
An abort error usually occurs when a il2cpp/c# exception occurs.

You can try to see what's happening via
```ts
Il2Cpp.installExceptionListener("all");
```
Before you invoke the method

**Reply by vfsfitvnm** [reply]
PS - I know the game you are playing with. I remember I got some abort errors when I tried to invoke some methods from the Frida thread. You can try to execute it in the main one via:
```ts
Il2Cpp.scheduleOnInitializerThread(() => {
    // ...
});
```

Keep in mind the implementstion is very poor and it might not work - however you get the concept - you have to invoke that method from the correct thread

**Reply by SamueleD78** [reply]
Thank you! That did the trick! Actually the call didn't give anything back, but that could be because there isn't anything.
At least i don't have anymore the error.

You were really gentle, please excuse me for this last late reply, but i had a nightmare of a weekend ^^

**Reply by vfsfitvnm** [reply]
> Actually the call didn't give anything back

I suspect it might be because the implementation is still crappy - also, the game must be to the foreground in order to execute

**Reply by SamueleD78** [reply]
if was in foreground, so maybe this means the implementation it's still not working.

Thank you anyway for everything you have done, there a lot of things working.

Do you think with android could it work better?

**Reply by vfsfitvnm** [reply]
> if was in foreground, so maybe this means the implementation it's still not working.

[`scheduleOnInitializerThread`](https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/il2cpp/base.ts#L183) works by intercepting `il2cpp_thread_current`, which it seemed to be constantly called by the main Il2Cpp thread; however this not the case for some applications.

So, the idea is to intercept a function (a native / il2cpp one) you know is being frequently called from that thread, e.g. using `Il2Cpp.Method.implementation` will get the job done

**Reply by SamueleD78** [reply]
so just to be clear:

**instead** of using `scheduleOnInitializerThread`, i have to search a function called often, hook the implementation, run my code into it because this way it's executed in the proper thread.

so my code atm is:

```
Il2Cpp.scheduleOnInitializerThread(() => {
    // my code with the invoke returing null atm goes here
});
```

it'll become:

```
Il2Cpp.perform(() => {
    const mscorlib = Il2Cpp.Domain.assembly("Assembly-CSharp").image;
    const method = mscorlib.class("someclass").method<T>("somemethod")
    
    method.implementation = function (someparam: T) {
        // my code with the invoke returing null instead goes here
        
        return this.method<T>("somemethod").invoke(someparam); // i'm just using the function for being in the thread space, i don't care what it does
    };
});
```

**Reply by vfsfitvnm** [reply]
Correct. 
You may take a look at `seen` (it is what I used, don't know if it still exists). Please take note, so I can obscure it.

However, this would be more convenient (not tested):
```ts
Reflect.defineProperty(Il2Cpp, "scheduleOnInitializerThread", {
    value: function <T>(block: () => T | Promise<T>): Promise<T> {
        const AssemblyCSharp = Il2Cpp.Domain.assembly("Assembly-CSharp").image;
        const Target = AssemblyCSharp.class("...").method("...");

        return new Promise<T>(resolve => {
            const listener = Interceptor.attach(Target.virtualAddress, () => {
                listener.detach();
                const result = block();
                setImmediate(() => resolve(result));
            });
        });
    }
});
```
